### PR TITLE
53 minimum marketable product   challenge completion handling

### PR DESCRIPTION
**Pull request by RadiationOcelot**
Updates the database whenever a HTTP request is made. Checks whether a challenge is finished, and if so, updates the database accordingly, and adds points to the winning user.

Added a check to block attempts if the challenge has already finished. Currently redirects them back to viewing the challenge, but maybe there's a better way to communicate that they had run out of time?

Also allows anonymous users to see all submissions for a challenge once the challenge has finished.

### CODE REVIEW

File: app/database.py [hunk 1]
```diff
@@ -57,4 +57,29 @@ def createNewChallenge(creator_id, title, path, dt_submit, dt_finish):
...
+    winningSubmission = None
+    for submission in challenge.attempts:
+        # If there's a tie, the earliest submission should be the winner, and in the database, it automatically sorts via attempt submission
+        if submission.article_no < maxArticles:
+            maxArticles = submission.article_no
+            winningSubmission = submission.id
+    return winningSubmission
+
+# Query the database to see if a challenge is finished. If so, update the database to store who won the challenge.
+def updateDB():
```
Comment (DeltaO3) [inline app/database.py:74]: Minor nitpick bu i think a different name for the function would be better? updateDB implies a lot of things, while you are just checking if challenges are finished?

File: app/database.py [hunk 2]
```diff
@@ -57,4 +57,29 @@ def createNewChallenge(creator_id, title, path, dt_submit, dt_finish):
...
+    challenges = Challenge.query.all()
+    for challenge in challenges:
+        # challenge.finished == False is most likely bad practice
+        if ((challenge.dt_finish-int(time.time())) <= 0 and challenge.finished == False):
```
Comment (DeltaO3) [inline app/database.py:78]: is challenge.finished necessary here, even? could we not remove it from the table entirely?

Comment (DeltaO3) [reply app/database.py:78]: Although i see why (it could slow things down with many database calls). I'm fine to keep it like this for now and just make a note to look over it later

File: app/routes.py [hunk 3]
```diff
@@ -46,9 +50,13 @@ def create():
 #Submit should always include an id parameter
 @app.route('/submit', methods=['POST'])
 def submit():
+    updateDB()
     submitForm = SubmitForm()
     form = LoginForm()
     challenge = getChallenge(int(submitForm.challenge_id.data)).toDict()
+    if challenge.finished:
+        # There has to be a better way to communicate that the time has run out other than reloading the page
```
Comment (DeltaO3) [inline app/routes.py:58]: Agree with comment, but fine with this for now

File: app/routes.py [hunk 4]
```diff
@@ -22,13 +23,16 @@ def index():
 #Create challenge page
 @app.route('/create', methods=['GET', 'POST'])
 def create():
+    updateDB()
```
Comment (DeltaO3) [inline app/routes.py:26]: Do we need to updateDB in create? and most places that arent directly related to an existing challenge (creating account, logging in, logging out?)

**Review by DeltaO3 — APPROVED** [review]
Looks pretty good, a couple of minor comments that I think should be addressed (but I dont think changes are necessarily required right now.) approving, but read comments first and decide if you want to make changes.

File: app/database.py (hunk 1 above)
Comment (RadiationOcelot) [reply app/database.py:74]: While making the code I couldn't think of a better function name. Maybe something like `checkChallenges()` or `checkChallengesCompleted()` ?

File: app/database.py (hunk 2 above)
Comment (RadiationOcelot) [reply app/database.py:78]: When I made the database I thought it would be much simpler for us in the future if we only had to check `challenge.finished` rather than `challenge.dt_finish-int(time.time())`

File: app/routes.py (hunk 4 above)
Comment (RadiationOcelot) [reply app/routes.py:26]: Ah, probably not, I'll remove it there.

File: app/database.py (hunk 1 above)
Comment (DeltaO3) [reply app/database.py:74]: checkChallengesCompleted() is better imo, if a bit wordy

File: app/database.py (hunk 2 above)
Comment (DeltaO3) [reply app/database.py:78]: Yeah, definitely agree there.

File: app/database.py [hunk 5]
```diff
@@ -57,4 +57,29 @@ def createNewChallenge(creator_id, title, path, dt_submit, dt_finish):
         dt_finish=dt_finish,
         finished=False)
     db.session.add(newChallenge)
-    db.session.commit()
\ No newline at end of file
+    db.session.commit()
+
+# Given a Challenge model, find the submission with the fewest articles in its path (the winner of the challenge)
+def findWinner(challenge):
+    maxArticles = 51 # For what we currently have, each path has a maximum length of 50 articles
```
Comment (RedBlueCarrots) [inline app/database.py:64]: Although maximum is currently 50, apart from the field list everything is set up to be dynamically scaled, theoretically beyond 50.

I think a better solution would be to set maxArticles to -1, and then in the below if statement add an "or maxArticles == -1"

File: app/routes.py (hunk 3 above)
Comment (RedBlueCarrots) [reply app/routes.py:57]: ^

File: app/utilities.py [hunk 6]
```diff
@@ -15,6 +15,8 @@ def secondsToTime(totalSeconds):
 	    timeString += " "+str(minutesLeft)+" minute"
 	    if minutesLeft > 1:
 	        timeString += "s"
+	if timeString == "":
+	    timeString += "just about to finish"
```
Comment (RedBlueCarrots) [inline app/utilities.py:19]: Capitalise pls

**Review by RedBlueCarrots — CHANGES_REQUESTED** [review]
Couple of smol changes I think would be beneficial, but very good apart from that

File: app/database.py (hunk 5 above)
Comment (RadiationOcelot) [reply app/database.py]: Done

File: app/utilities.py (hunk 6 above)
Comment (RadiationOcelot) [reply app/utilities.py]: Done

**Review by RedBlueCarrots — APPROVED** [review]
Lgtm
